frame/staking/src/lib.rs

```diff
@@ -1371,7 +1751,185 @@ impl<T: Trait> Module<T> {
+		if let Some(queued_score) = Self::queued_score() {
+			ensure!(
+				is_score_better(queued_score, claimed_score),
+				Error::<T>::PhragmenWeakSubmission,
+			)
+		}
+
+		// Check that the number of presented winners is sane. Most often we have more candidates
+		// that we need. Then it should be Self::validator_count(). Else it should be all the
+		// candidates.
+		let snapshot_length = <SnapshotValidators<T>>::decode_len()
+			.map_err(|_| Error::<T>::SnapshotUnavailable)?;
+		let desired_winners = Self::validator_count().min(snapshot_length as u32);
+		ensure!(winners.len() as u32 == desired_winners, Error::<T>::PhragmenBogusWinnerCount);
+
+		// decode snapshot validators.
+		let snapshot_validators = Self::snapshot_validators()
+			.ok_or(Error::<T>::SnapshotUnavailable)?;
+
+		// check if all winners were legit; this is rather cheap. Replace with accountId.
+		let winners = winners.into_iter().map(|widx| {
+			snapshot_validators.get(widx as usize).cloned().ok_or(Error::<T>::PhragmenBogusWinner)
+		}).collect::<Result<Vec<T::AccountId>, Error<T>>>()?;
+
+		// decode the rest of the snapshot.
+		let snapshot_nominators = <Module<T>>::snapshot_nominators()
+			.ok_or(Error::<T>::SnapshotUnavailable)?;
+
+		// helpers
+		let nominator_at = |i: NominatorIndex| -> Option<T::AccountId> {
+			// NOTE: we assume both `ValidatorIndex` and `NominatorIndex` are smaller than usize.
+			snapshot_nominators.get(i as usize).cloned()
+		};
+		let validator_at = |i: ValidatorIndex| -> Option<T::AccountId> {
+			// NOTE: we assume both `ValidatorIndex` and `NominatorIndex` are smaller than usize.
+			snapshot_validators.get(i as usize).cloned()
+		};
+		let slashable_balance = |w: &T::AccountId| -> ExtendedBalance {
+			<T::CurrencyToVote as Convert<BalanceOf<T>, u64>>::convert(
+				Self::slashable_balance_of(w)
+			) as ExtendedBalance
+		};
+
+		// un-compact.
+		let assignments = compact_assignments.into_assignment(
+			nominator_at,
+			validator_at,
+		).map_err(|_| Error::<T>::PhragmenBogusCompact)?;
+
+		// check all nominators actually including the claimed vote. Also check correct self votes.
+		// Note that we assume all validators and nominators in `assignments` are properly bonded,
+		// because they are coming from the snapshot via a given index.
+		for Assignment { who, distribution } in assignments.iter() {
+			let is_validator = <Validators<T>>::contains_key(&who);
+			let maybe_nomination = Self::nominators(&who);
+
+			if !(maybe_nomination.is_some() ^ is_validator) {
+				// all of the indices must map to either a validator or a nominator. If this is ever
+				// not the case, then the locking system of staking is most likely faulty, or we
+				// have bigger problems.
+				debug::native::error!(
+					target: "staking",
+					"detected an error in the staking locking and snapshot."
+				);
+				// abort.
+				return Err(Error::<T>::PhragmenBogusNominator);
```

Comment:
This probably deserves a panic.. I don't even have a test case to show it.

Reply:
If I read correctly we accept ppl to chill during the electionStatus::Open period, thus a staker can chill and a solution proposed containing him will raise this error

If my analyse is correct then I would remove the debug::native::error as it is not an error in case we don't find nominators and validators,

maybe the best would be to make a full match:
```rust
match (validator, maybe_nomination) {
(false, false) => return Err(PhragmenBogusNominator),
(true, false) => do validator case
(false, true) => do nominator case
(true, true) => return Err(StakingInternalError),
}
```

Reply:
> If I read correctly we accept ppl to chill during the electionStatus::Open period, thus a staker can chill and a solution proposed containing him will raise this error

No, we actually disallow any change to the staking status. See the Signed Extension at the bottom. 

I would hold on about this since I am waiting for @rphmeier to get back to me about slashing and how it should fit into offchain Phragmen. Within the scope of this PR, the if is not even needed since we have the locking. Taking slashing into account, some validators might get automatically chilled in the duration of the election window. Still, not much that we can do about it, but maybe good to have some logs about it.

Reply:
oh great 👍

Reply:
Just an update here that we now queue all offences at the offences module, no validator may get chilled by the system either at this point.